File: Haskell/Euler3.hs
--returns list of all prime factors of x
primeFactors :: Int -> [Int]
primeFactors x = primeFactors' x 2

primeFactors' :: Int -> Int -> [Int]
primeFactors' x f
    | x<=1                           = []                               --reached 1: all factors found
    | ((mod x f)==0) && (isPrime f)  = f:(primeFactors' (div x f) 2)    --prime factor found, divide and find new factors
    | otherwise                      = primeFactors' x (f+1)            --not a factor, check next



--returns true if x is prime
isPrime :: Int -> Bool
isPrime x = isPrime' x 2

isPrime' :: Int -> Int -> Bool
isPrime' num f
    | num==f           = True                   --num reached before factor found: prime
    | (mod num f)==0   = False                  --factor found: not prime
    | otherwise        = isPrime' num (f+1)     --not a factor: check next f

File: Python/Euler3.py
import math


def is_prime(n):
    if n > 2 and n % 2 == 0:
        return False
    for i in range(2, n // 2):
        if n % i == 0:
            return False
    return True


def factors(n):
    candidates = range(2, math.ceil((math.sqrt(n))))
    f1 = list(filter(lambda x: n % x == 0, candidates))
    f2 = list(map(lambda x: n // x, f1))
    f1.extend(f2)
    return f1


def prime_factors(n):
    return list(filter(lambda x: is_prime(x), factors(n)))


print(prime_factors(600851475143))
